ash/src/vk/prelude.rs

```diff
@@ -63,6 +63,15 @@ pub unsafe trait TaggedStructure {
     const STRUCTURE_TYPE: vk::StructureType;
 }
 
+/// Implemented for every structure that extends base structure `B`. Concretely that means struct
+/// `B` is listed in its array of [`structextends` in the Vulkan registry][1].
+///
+/// Similar to [`TaggedStructure`] , all `unsafe` implementers of this trait must guarantee that
+/// their structure is layout-compatible [`vk::BaseInStructure`] and [`vk::BaseOutStructure`].
+///
+/// [1]: https://registry.khronos.org/vulkan/specs/latest/styleguide.html#extensions-interactions
+pub unsafe trait StructExtends<B> {}
```

Comment:
I would've liked to `: TaggedStructure` to imply these docs, but that makes the trait no longer object-safe meaning that folks cannot pass `&mut dyn StructExtends<B>` into any of our `.push*()`/`.extend()` functions.

If we go that route we can also drop `+ ?Sized` from the trait bounds and put `: Sized` on `StructExtends` directly :)

Reply:
We can achieve the best of both worlds by splitting `TaggedStructure` into two traits: one that's object-safe and implies the contract (maybe `AnyTaggedStructure`), and one that inherits that trait (maybe `TaggedStructure: AnyTaggedStructure`), and adds `const STRUCTURE_TYPE`.

Bit of ceremony for a really obscure use case, but it doesn't seem like it'd hurt.

Reply:
Let's do that when we get there? I'm not exactly set on keeping the ability to pass `&dyn` around, but folks like #953 seem to be wanting that and even added a `ASH_DYNAMIC` to get around this, which I think we should avert by splitting the trait like you described, and not having a `const` available for "generic" unknown objects.

Having to implement two (hierarchical) traits just to share the layout requirement documentation sounds like overdoing it.

Reply:
When I suggested to add our `push()` functions to this not-object-safe `TaggedStructure` in https://github.com/ash-rs/ash/issues/879, I think we'd actually break that `&dyn` case already without the `AnyTaggedStructure` (or `RuntimeTaggedStructure`)? I think we'll add it in the next PR then.

Reply:
> Having to implement two (hierarchical) traits

We could flip that around with `impl<T: TaggedStructure> AnyTaggedStructure for T`; then it's completely automatic.

Reply:
That's great, I didn't expect that to work because of a circular `TaggedStructure: AnyTaggedStructure` dependency, but it does :tada: